src/OpenTelemetry.Exporter.Geneva/Internal/TableNameSerializer.cs

```diff
@@ -0,0 +1,313 @@
+            // it will be converted to the corresponding upper-case.
+            buffer[cursor++] = (byte)(firstChar - 32);
+            ++validNameLength;
+        }
+        else
+        {
+            // Not a valid name.
+            return 0;
+        }
+
+        for (int i = 1; i < categoryName.Length; ++i)
+        {
+            if (validNameLength == MaxSanitizedEventNameLength)
+            {
+                break;
+            }
+
+            var cur = categoryName[i];
+            if ((cur >= 'a' && cur <= 'z') || (cur >= 'A' && cur <= 'Z') || (cur >= '0' && cur <= '9'))
+            {
+                buffer[cursor++] = (byte)cur;
+                ++validNameLength;
+            }
+        }
+
+        // Backfilling MessagePack serialization protocol and valid category length to the startIdx of the categoryName byte array.
+        MessagePackSerializer.WriteStr8Header(buffer, 0, validNameLength);
+
+        return cursor;
+    }
+
+    private byte[] ResolveTableMappingForCategoryName(string categoryName)
+    {
+        var tableNameCache = this.m_tableNameCache;
+
+        if (tableNameCache.TryGetValue(categoryName, out byte[] tableName))
+        {
+            return tableName;
+        }
+
+        return this.ResolveTableMappingForCategoryNameRare(categoryName);
+    }
+
+    private byte[] ResolveTableMappingForCategoryNameRare(string categoryName)
+    {
+        lock (this.m_lockObject)
+        {
+            var tableNameCache = this.m_tableNameCache;
+
+            // Check if another thread added the mapping while we waited on the
+            // lock.
+            if (tableNameCache.TryGetValue(categoryName, out byte[] tableName))
+            {
+                return tableName;
+            }
+
+            byte[] mappedTableName = null;
+
+            // If user configured table name mappings run resolution logic.
+            if (this.m_tableMappings != null
+                && !this.m_tableMappings.TryGetValue(categoryName, out mappedTableName))
+            {
+                // Find best match if an exact match was not found.
+
+                string currentKey = null;
+
+                foreach (var mapping in this.m_tableMappings)
+                {
+                    if (!categoryName.StartsWith(mapping.Key, StringComparison.OrdinalIgnoreCase))
+                    {
+                        continue;
+                    }
+
+                    if (currentKey == null || mapping.Key.Length >= currentKey.Length)
+                    {
+                        currentKey = mapping.Key;
+                        mappedTableName = mapping.Value;
+                    }
+                }
+            }
+
+            mappedTableName ??= !this.m_shouldPassThruTableMappings
+                ? this.m_defaultTableName
+                : s_passthroughTableName;
+
+            if (mappedTableName.Length <= 0 && tableNameCache.CachedSanitizedTableNameCount < MaxCachedSanitizedTableNames)
```

Comment:
Why does this limit of 1024 matter? If I understand correctly, regardless of the `CachedSanitizedTableNameCount` value, we would always add a new category as a key to the `m_tableNameCache`. So why not just add the actual sanitized category name as the value instead of `s_passthroughTableName` when the 1024 limit is reached.

Reply:
Short answer: Today we don't cache any of the sanitized names so this was me being conservative. 

Long answer:

> If I understand correctly, regardless of the CachedSanitizedTableNameCount value, we would always add a new category as a key to the m_tableNameCache

Correct

> So why not just add the actual sanitized category name as the value instead of s_passthroughTableName when the 1024 limit is reached.

Just to make it less greedy. Each category added after 1024 will grow by a pointer to s_passthroughTableName instead of a pointer to some freshly allocated buffer (up to 52 bytes IIRC).

Reply:
> Just to make it less greedy. Each category added after 1024 will grow by a pointer to s_passthroughTableName instead of a pointer to some freshly allocated buffer (up to 52 bytes IIRC).

Would 52 bytes be considered high? For every new set of 1024 keys, it's only 52 KB max.

Reply:
Honestly, didn't think too much about it. Just nice to prevent unconstrained growth where we can. The check is only processed the first time we see a category name value and then we cache the result. So it is pretty inexpensive (a few nanos) to have the cap. You want me to remove it?

Reply:
More than the `if` check I was thinking we could save on another round of parsing of Category name when we have more than 1024 category names.

It will come down to the cost of copying a small `byte[]` to the buffer vs executing WriteSanitizedCategoryName (parsing + serialize). I was wondering if we anyway have to parse the string for dictionary lookup why not just get hold of the `byte[]` instead of again parsing and serializing the sanitized Category name.

Reply:
Hmm. I think there is some confusion about what is going on. So we always cache _something_. For a given category name the cache will be one of these things...

1) A defined mapping (pointer to something in m_tableMappings).
2) The default table name (pointer to m_defaultTableName).
3) A sanitized category name (allocation).
4) Array.Empty<byte> if category name could not be sanitized (pointer to static).
5) The pass-through indicator if the cache was full (pointer to s_passthroughTableName).

The logic is...

* Lookup the mapping using category name from the cache.
* If the mapping is the pass-through indicator, write out the sanitized table name directly to the buffer.

For pass-through today the logic is...

* Check for a mapping using category name.
* If nothing was found and pass-through is enabled, write out the sanitized table name directly to the buffer.

So basically when the cache fills up we don't do double-work, we just fallback to the behavior we have today.

Reply:
Update... I synced up with @utpilla offline. The concern here is that in computing hash code for CategoryName in order to look into the cache "parses" the string. I have some ideas to improve this, but I don't thing we need to do anything here. The perf should be better for most, slightly worse for some.

* For users using at least 1 mapping, there is already a hash code computation + dictionary lookup in their path today. What these users will see: There should be see a speed-up for <= 1024 categories. Same speed as today for > 1024 categories.

* For users NOT using any mappings today, there is now a new hash code computation + dictionary lookup in their path. What these users will see: When using default table name, there should be a speed-up because we now copy a pre-built array using span helpers instead of calling into `MessagePackSerializer.SerializeAsciiString` each time. When using pass-through: Speed-up for <= 1024 categories. A slight slow-down for > 1024 categories.